In [1]:
# This Python 3 environment comes with many helpful analytics libraries installed
# It is defined by the kaggle/python Docker image: https://github.com/kaggle/docker-python
# For example, here's several helpful packages to load

import numpy as np # linear algebra
import pandas as pd # data processing, CSV file I/O (e.g. pd.read_csv)

# Input data files are available in the read-only "../input/" directory
# For example, running this (by clicking run or pressing Shift+Enter) will list all files under the input directory

import os
for dirname, _, filenames in os.walk('/kaggle/input'):
    for filename in filenames:
        print(os.path.join(dirname, filename))

# You can write up to 20GB to the current directory (/kaggle/working/) that gets preserved as output when you create a version using "Save & Run All" 
# You can also write temporary files to /kaggle/temp/, but they won't be saved outside of the current session

In [2]:
%%html
<video controls autoplay><source src="https://huggingface.co/sb3/ppo-LunarLander-v2/resolve/main/replay.mp4" type="video/mp4"></video>

In [3]:
!apt install swig cmake

Reading package lists... Done
Building dependency tree... Done
Reading state information... Done
cmake is already the newest version (3.22.1-1ubuntu1.22.04.1).
Suggested packages:
  swig-doc swig-examples swig4.0-examples swig4.0-doc
The following NEW packages will be installed:
  swig swig4.0
0 upgraded, 2 newly installed, 0 to remove and 46 not upgraded.
Need to get 1116 kB of archives.
After this operation, 5542 kB of additional disk space will be used.
Get:1 http://archive.ubuntu.com/ubuntu jammy/universe amd64 swig4.0 amd64 4.0.2-1ubuntu1 [1110 kB]
Get:2 http://archive.ubuntu.com/ubuntu jammy/universe amd64 swig all 4.0.2-1ubuntu1 [5632 B]
Fetched 1116 kB in 0s (3593 kB/s)0m

78Selecting previously unselected package swig4.0.
(Reading database ... 114840 files and directories currently installed.)
Preparing to unpack .../swig4.0_4.0.2-1ubuntu1_amd64.deb ...
7Progress: [  0%] [..........................................................] 87Progress: [ 11%] [######...............

In [29]:
!pip install moviepy

     ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 388.3/388.3 kB 6.0 MB/s eta 0:00:00a 0:00:01
  Preparing metadata (setup.py) ... done
  Obtaining dependency information for imageio_ffmpeg>=0.2.0 from https://files.pythonhosted.org/packages/1a/98/3df1d8dd8f2c121b6c588b1e0d604f36592d56df9c41fb155ed546c6a5ed/imageio_ffmpeg-0.4.9-py3-none-manylinux2010_x86_64.whl.metadata
   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 26.9/26.9 MB 48.0 MB/s eta 0:00:00:00:0100:01
  Created wheel for moviepy: filename=moviepy-1.0.3-py3-none-any.whl size=110721 sha256=c52056203c9106f4f41affe400f9c33664b8f48b1ec23cfa0cae49a1af78c6e9
  Stored in directory: /root/.cache/pip/wheels/96/32/2d/e10123bd88fbfc02fed53cc18c80a171d3c87479ed845fa7c1
Successfully built moviepy
  Attempting uninstall: decorator
    Found existing installation: decorator 5.1.1
    Uninstalling decorator-5.1.1:
      Successfully uninstalled decorator-5.1.1
ERROR: pip's dependency resolver does not currently take into account all the packages that

In [4]:
#Install dependencies
!pip install -r https://raw.githubusercontent.com/huggingface/deep-rl-class/main/notebooks/unit1/requirements-unit1.txt

     ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 177.5/177.5 kB 4.0 MB/s eta 0:00:00a 0:00:01
  Obtaining dependency information for swig from https://files.pythonhosted.org/packages/50/ce/d683fd61491983749484105a752291fe094d6247e336de1b76eab5600e14/swig-4.1.1.post0-py2.py3-none-manylinux_2_5_x86_64.manylinux1_x86_64.whl.metadata
  Obtaining dependency information for huggingface_sb3 from https://files.pythonhosted.org/packages/b5/2c/e8c25087c6e43e5e68659f2a4651988831d4b3e70791c433537c90d147ce/huggingface_sb3-3.0-py3-none-any.whl.metadata
     ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 925.5/925.5 kB 25.2 MB/s eta 0:00:0000:01
INFO: pip is looking at multiple versions of gymnasium[box2d] to determine which version is compatible with other requirements. This could take a while.
  Obtaining dependency information for gymnasium[box2d] from https://files.pythonhosted.org/packages/a8/4d/3cbfd81ed84db450dbe73a89afcd8bc405273918415649ac6683356afe92/gymnasium-0.29.1-py3-none-any.whl.metadata
  Obtaini

In [5]:
#During the notebook, we’ll need to generate a replay video. To do so, with colab, we need to have a virtual screen to be able to render the environment (and thus record the frames).
!sudo apt-get update
!apt install python-opengl
!apt install ffmpeg
!apt install xvfb
!pip3 install pyvirtualdisplay

Get:1 https://packages.cloud.google.com/apt cloud-sdk InRelease [6361 B]
Get:2 https://packages.cloud.google.com/apt google-fast-socket InRelease [5015 B]
Hit:3 http://archive.ubuntu.com/ubuntu jammy InRelease                         
Hit:4 http://packages.cloud.google.com/apt gcsfuse-focal InRelease             
Get:5 http://archive.ubuntu.com/ubuntu jammy-updates InRelease [119 kB]        
Get:6 https://developer.download.nvidia.com/compute/cuda/repos/ubuntu2204/x86_64  InRelease [1581 B]
Get:7 https://packages.cloud.google.com/apt cloud-sdk/main amd64 Packages [553 kB]
Get:8 http://security.ubuntu.com/ubuntu jammy-security InRelease [110 kB]      
Get:9 http://archive.ubuntu.com/ubuntu jammy-backports InRelease [109 kB]      
Get:10 https://developer.download.nvidia.com/compute/cuda/repos/ubuntu2204/x86_64  Packages [631 kB]
Get:11 http://archive.ubuntu.com/ubuntu jammy-updates/multiverse amd64 Packages [49.8 kB]
Get:12 http://archive.ubuntu.com/ubuntu jammy-updates/main amd64 Packa

In [ ]:
import os

os.kill(os.getpid(), 9)

In [1]:
# Virtual display
from pyvirtualdisplay import Display

virtual_display = Display(visible=0, size=(1400, 900))
virtual_display.start()

In [26]:
import gymnasium as gym
from huggingface_sb3 import load_from_hub, package_to_hub
from huggingface_hub import (
    notebook_login,
)
from stable_baselines3 import PPO
from stable_baselines3.common.env_util import make_vec_env
from stable_baselines3.common.vec_env import DummyVecEnv
from stable_baselines3.common.evaluation import evaluate_policy
from stable_baselines3.common.monitor import Monitor

In [6]:
# First, we create our environment called LunarLander-v2
env = gym.make('LunarLander-v2')

In [7]:
#Reset to initial state
observation, info = env.reset()

In [17]:
for _ in range(20):
    #Take random action
    action = env.action_space.sample()
    print('Action taken:', action)
    
    # Do this action in the environment and get
    # next_state, reward, terminated, truncated and info
    observation, reward, terminated, truncated, info = env.step(action)
    
    # If the game is terminated (in our case we land, crashed) or truncated (timeout)
    if terminated or truncated:
        print('Environment is reset')
        observation, info = env.reset()
env.close()

Action taken: 3
Action taken: 2
Action taken: 3
Action taken: 2
Action taken: 3
Action taken: 0
Action taken: 2
Action taken: 2
Action taken: 2
Action taken: 1
Action taken: 1
Action taken: 2
Action taken: 1
Action taken: 1
Action taken: 2
Action taken: 0
Action taken: 2
Action taken: 2
Action taken: 3
Action taken: 2


In [18]:
# We create our environment with gym.make("<name_of_the_environment>")
env = gym.make("LunarLander-v2")
env.reset()
print("_____OBSERVATION SPACE_____ \n")
print("Observation Space Shape", env.observation_space.shape)
print("Sample observation", env.observation_space.sample())

_____OBSERVATION SPACE_____ 

Observation Space Shape (8,)
Sample observation [-35.911915    16.929241    -0.48472762  -3.5126724    2.9744513
   1.4508443    0.3582742    0.17189333]


We see with Observation Space Shape (8,) that the observation is a vector of size 8, where each value contains different information about the lander:

* Horizontal pad coordinate (x)
* Vertical pad coordinate (y)
* Horizontal speed (x)
* Vertical speed (y)
* Angle
* Angular speed
* If the left leg contact point has touched the land (boolean)
* If the right leg contact point has touched the land (boolean)

In [19]:
print("\n _____ACTION SPACE_____ \n")
print("Action Space Shape", env.action_space.n)
print("Action Space Sample", env.action_space.sample())  # Take a random action


 _____ACTION SPACE_____ 

Action Space Shape 4
Action Space Sample 1


The action space (the set of possible actions the agent can take) is discrete with 4 actions available 🎮:

* Action 0: Do nothing,
* Action 1: Fire left orientation engine,
* Action 2: Fire the main engine,
* Action 3: Fire right orientation engine.


After every step a reward is granted. The total reward of an episode is the sum of the rewards for all the steps within that episode.

For each step, the reward:

* Is increased/decreased the closer/further the lander is to the landing pad.
* Is increased/decreased the slower/faster the lander is moving.
* Is decreased the more the lander is tilted (angle not horizontal).
* Is increased by 10 points for each leg that is in contact with the ground.
* Is decreased by 0.03 points each frame a side engine is firing.
* Is decreased by 0.3 points each frame the main engine is firing.

The episode receive an additional reward of -100 or +100 points for crashing or landing safely respectively.

An episode is considered a solution if it scores at least 200 points.

# Vectorized Environment
We create a vectorized environment (a method for stacking multiple independent environments into a single environment) of 16 environments, this way, we’ll have more diverse experiences during the training.

In [20]:
# Create the environment
env = make_vec_env("LunarLander-v2", n_envs=16)

In [21]:
# Instantiate the agent
# model = PPO('MlpPolicy', env, verbose=1)
model = PPO(
    policy="MlpPolicy",
    env=env,
    n_steps=1024,
    batch_size=64,
    n_epochs=4,
    gamma=0.999,
    gae_lambda=0.98,
    ent_coef=0.01,
    verbose=1,
)

# Train the agent
model.learn(total_timesteps=1000000)

# Save the model
model_name = "ppo-LunarLander-v2"
model.save(model_name)

Using cuda device
---------------------------------
| rollout/           |          |
|    ep_len_mean     | 89.7     |
|    ep_rew_mean     | -175     |
| time/              |          |
|    fps             | 2397     |
|    iterations      | 1        |
|    time_elapsed    | 6        |
|    total_timesteps | 16384    |
---------------------------------
------------------------------------------
| rollout/                |              |
|    ep_len_mean          | 95.1         |
|    ep_rew_mean          | -165         |
| time/                   |              |
|    fps                  | 2052         |
|    iterations           | 2            |
|    time_elapsed         | 15           |
|    total_timesteps      | 32768        |
| train/                  |              |
|    approx_kl            | 0.0057260757 |
|    clip_fraction        | 0.0165       |
|    clip_range           | 0.2          |
|    entropy_loss         | -1.38        |
|    explained_variance   | -0.00249    

# Evaluate the agent 

In [22]:
# @title
eval_env = Monitor(gym.make("LunarLander-v2"))
mean_reward, std_reward = evaluate_policy(model, eval_env, n_eval_episodes=10, deterministic=True)
print(f"mean_reward={mean_reward:.2f} +/- {std_reward}")

mean_reward=260.89 +/- 19.449573305716058


So we got mean reward of 260.89, which means our lander is ready to land on moon

# Push to huggingface model hub
- Need to create write token to push the model

In [23]:
notebook_login()
!git config --global credential.helper store

In [24]:
# Define the name of the environment
env_id = "LunarLander-v2"
#Define the model architecture we used
model_architecture = "PPO"
## Define a repo_id
## repo_id is the id of the model repository from the Hugging Face Hub (repo_id = {organization}/{repo_name} for instance ThomasSimonini/ppo-LunarLander-v2
repo_id = "mlwithrakesh/ppo-LunarLander-v2"
## Define the commit message
commit_message = "Upload PPO LunarLander-v2 trained agent"

In [27]:
# Create the evaluation env and set the render_mode="rgb_array"
eval_env = DummyVecEnv([lambda: Monitor(gym.make(env_id, render_mode="rgb_array"))])

In [31]:
package_to_hub(
    model=model,  # Our trained model
    model_name=model_name,  # The name of our trained model
    model_architecture=model_architecture,  # The model architecture we used: in our case PPO
    env_id=env_id,  # Name of the environment
    eval_env=eval_env,  # Evaluation Environment
    repo_id=repo_id,  # id of the model repository from the Hugging Face Hub (repo_id = {organization}/{repo_name} for instance ThomasSimonini/ppo-LunarLander-v2
    commit_message=commit_message,
)

ℹ This function will save, evaluate, generate a video of your agent,
create a model card and push everything to the hub. It might take up to 1min.
This is a work in progress: if you encounter a bug, please open an issue.


Exception ignored in: <function VecVideoRecorder.__del__ at 0x7c89a0342b00>
Traceback (most recent call last):
  File "/opt/conda/lib/python3.10/site-packages/stable_baselines3/common/vec_env/vec_video_recorder.py", line 113, in __del__
    self.close_video_recorder()
  File "/opt/conda/lib/python3.10/site-packages/stable_baselines3/common/vec_env/vec_video_recorder.py", line 104, in close_video_recorder
    self.video_recorder.close()
  File "/opt/conda/lib/python3.10/site-packages/gymnasium/wrappers/monitoring/video_recorder.py", line 161, in close
    clip.write_videofile(self.path, logger=moviepy_logger)
  File "/opt/conda/lib/python3.10/site-packages/decorator.py", line 232, in fun
    for i, extra in enumerate(extras):
  File "/opt/conda/lib/python3.10/site-packages/moviepy/decorators.py", line 54, in requires_duration
    return f(clip, *a, **k)
  File "/opt/conda/lib/python3.10/site-packages/decorator.py", line 232, in fun
    for i, extra in enumerate(extras):
  File "/opt/con

Saving video to /tmp/tmp5xnh5glo/-step-0-to-step-1000.mp4
Moviepy - Building video /tmp/tmp5xnh5glo/-step-0-to-step-1000.mp4.
Moviepy - Writing video /tmp/tmp5xnh5glo/-step-0-to-step-1000.mp4

✘ must be real number, not NoneType
✘ We are unable to generate a replay of your agent, the package_to_hub
process continues
✘ Please open an issue at
https://github.com/huggingface/huggingface_sb3/issues
Moviepy - Building video /tmp/tmp5xnh5glo/-step-0-to-step-1000.mp4.
Moviepy - Writing video /tmp/tmp5xnh5glo/-step-0-to-step-1000.mp4

ℹ Pushing repo mlwithrakesh/ppo-LunarLander-v2 to the Hugging Face
Hub


ppo-LunarLander-v2.zip:   0%|          | 0.00/147k [00:00<?, ?B/s]

ℹ Your model is pushed to the Hub. You can view your model here:
https://huggingface.co/mlwithrakesh/ppo-LunarLander-v2/tree/main/


'https://huggingface.co/mlwithrakesh/ppo-LunarLander-v2/tree/main/'

# Load a saved LunarLander model from the Hub

Because the model I download from the Hub was trained with Gym (the former version of Gymnasium) we need to install shimmy a API conversion tool that will help us to run the environment correctly.

Shimmy Documentation: https://github.com/Farama-Foundation/Shimmy

In [33]:
!pip install shimmy

  Obtaining dependency information for shimmy from https://files.pythonhosted.org/packages/dc/f9/07ef16463db14ac1b30f149c379760f5cacf3fc677b295d29a92f3127914/Shimmy-1.3.0-py3-none-any.whl.metadata


In [34]:
repo_id = "mlwithrakesh/ppo-LunarLander-v2"  # The repo_id
filename = "ppo-LunarLander-v2.zip"  # The model filename.zip

# When the model was trained on Python 3.8 the pickle protocol is 5
# But Python 3.6, 3.7 use protocol 4
# In order to get compatibility we need to:
# 1. Install pickle5 (we done it at the beginning of the colab)
# 2. Create a custom empty object we pass as parameter to PPO.load()
custom_objects = {
    "learning_rate": 0.0,
    "lr_schedule": lambda _: 0.0,
    "clip_range": lambda _: 0.0,
}

checkpoint = load_from_hub(repo_id, filename)
model = PPO.load(checkpoint, custom_objects=custom_objects, print_system_info=True)

== CURRENT SYSTEM INFO ==
- OS: Linux-5.15.133+-x86_64-with-glibc2.35 # 1 SMP Wed Nov 8 17:30:28 UTC 2023
- Python: 3.10.12
- Stable-Baselines3: 2.0.0a5
- PyTorch: 2.0.0
- GPU Enabled: True
- Numpy: 1.24.3
- Cloudpickle: 3.0.0
- Gymnasium: 0.28.1
- OpenAI Gym: 0.26.2

== SAVED MODEL SYSTEM INFO ==
- OS: Linux-5.15.133+-x86_64-with-glibc2.35 # 1 SMP Wed Nov 8 17:30:28 UTC 2023
- Python: 3.10.12
- Stable-Baselines3: 2.0.0a5
- PyTorch: 2.0.0
- GPU Enabled: True
- Numpy: 1.24.3
- Cloudpickle: 3.0.0
- Gymnasium: 0.28.1
- OpenAI Gym: 0.26.2



In [35]:
# @title
eval_env = Monitor(gym.make("LunarLander-v2"))
mean_reward, std_reward = evaluate_policy(model, eval_env, n_eval_episodes=10, deterministic=True)
print(f"mean_reward={mean_reward:.2f} +/- {std_reward}")

mean_reward=266.50 +/- 19.21535826587336
